accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,871 @@
+
+    The condition is inferred with a context type of `bool`.
+
+    The inferred list element type of `element` is the upper bound of `S1` and
+    `S2` where `S1` is the inferred list element type of `p1` and `S2` is the
+    inferred list element type of `p2`, both with element context type `P`.
+
+*   If `element` is a `forElement` with `element` `p1` then:
+
+    Inference for the iterated expression and the controlling variable is done
+    as for the corresponding `for` or `await for` statement.
+
+    The inferred list element type of `element` is the inferred list element
+    type of `p1` with element context type `P`.
+
+*Note: `element` cannot be a `mapEntry` as those are not allowed inside list
+literals.*
+
+Finally, we define inference on a `listLiteral` *collection* as follows:
+
+*   The downwards element context for inference of the elements of *collection*
+    is `P` where `P` is `T` if downwards inference constraints the type of
+    collection to `Iterable<T>` for some `T`. Otherwise, `P` is `?`.
+
+    *   If `P` is `?` then the static type of *collection* is `List<T>` where
+        `T` is the upper bound of the inferred list element types of the
+        elements.
+
+    *   Otherwise, the static type of *collection* is `List<T>` where `T` is
+        determined by downwards inference.
+
+### Type promotion
+
+An `ifElement` interacts with type promotion in the same way that `if`
+statements do. Given an `ifElement` with condition `condition`, then element
+`thenElement` and optional else element `elseElement`:
+
+*   If `condition` shows that a local variable *v* has type `T`, then the type
+    of *v* is known to be `T` in `thenElement`, unless any of the following are
+    true:
+
+    *   *v* is potentially mutated in `thenElement`,
+    *   *v* is potentially mutated within a function other than the one where
+        *v* is declared, or
+    *   *v* is accessed by a function defined in `thenElement` and
+    *   *v* is potentially mutated anywhere in the scope of *v*.
+
+*Note: The type promotion rules for `if` will likely get more sophisticated in a
+future version of Dart because of non-nullable types. When that happens, `if`
+elements should continue to match `if` statements.*
+
+### Compile-time errors
+
+After type inference and disambiguation, the collection is checked for other
+compile-time errors. It is a compile-time error if:
+
+*   The collection is a map literal and *leaf elements* contains any
+    `expressionElement` elements.
+
+    ```dart
+    <String, int>{"not entry"} // Error.
+    ```
+
+*   The collection is a set literal and *leaf elements* contains any `mapEntry`
+    elements.
+
+    ```dart
+    ["not": "expression"] // Error.
+    <String>{"not": "expression"} // Error.
+    ```
+
+    *We prohibit map entries in list literals syntacitcally, earlier in the
+    proposal.*
+
+*   The collection is a list and the type of any of the *leaf elements* may not
```

Comment:
A spread element is a leaf element. Have we defined the assignability of a spread element to an element type?

Reply:
A spread is a "leaf element" in terms of the AST, but not a *leaf element* as defined by the proposal.

Reply:
My bad. Got confused about what leaf elements are.